parsl/dataflow/task_status_poller.py

```diff
@@ -0,0 +1,62 @@
+import logging
+import parsl
+import time
+from typing import Dict, List, Sequence
+
+from parsl.dataflow.executor_status import ExecutorStatus
+from parsl.dataflow.strategy import Strategy
+from parsl.executors.base import ParslExecutor
+from parsl.providers.provider_base import JobStatus
+
+logger = logging.getLogger(__name__)
+
+
+class PollItem(ExecutorStatus):
```

Comment:
While this is fine and works great, a simpler approach would be to add a timed cache decorator in the provider, since the provider also holds the timeout period.

Reply:
So I gave this some considerable thought. But before I write about that, let me see if I understand you correctly. The following would help:
- do you mean discarding the centralized polling loop altogether or keeping the polling loop but having caching done on provider.getStatus()?
- do you have a specific caching decorator in mind, or did you mean that something can be written if needed?

Reply:
Central polling is better than each provider polling separately. 
Unfortunately `functools.lru_cache` doesn't have a timer for staleness, so we might have to do something like this : 

```python
import functools
import time

class TimedCache():
    def __init__(self, valid_time=2):                
        self.valid_time = valid_time
        print("valid_time = {}".format(valid_time))        

    def __call__(self, func):        
        print("here")        
        def wrapper(*args, **kwargs):
            
            if time.time() > self.cache_time + self.valid_time:
                print("Stale")
                self.cached_result = func(*args, **kwargs)
                self.cache_time = time.time()                
            else:
                print("Fetching cached")
                    
            return self.cached_result
        
        self.cache_time = 0
        self.cached_result = None
        return wrapper

@TimedCache(60)
def status( ...)
```

Reply:
by "central polling" I mean one polling loop for all consumers (strategy, error handling, etc.) vs. each of (strategy, error handling, etc.) calling executor.status() -> provider.status() on their own.

Anyway, the scheme in this PR is fundamentally single-threaded. There is one thread that maintains the status of the system and it dispatches that status to various consumers who deal with a snapshot of the status of the system. It's true that it requires some help from the providers in determining how often certain things should be done.

(1) We could keep the polling loop as it is, remove PollItem and get_polling_interval() methods, implement a cache as above, and have the polling loop call status() on all providers at some regular interval that is smaller than the smallest polling interval (this will be a somewhat magical decision, since the smallest polling interval would not be known to the loop). This is almost fine.

(2) We could also abandon the polling loop entirely and have the strategy and whatever other entities call status() on their own, knowing that the caching decorator does the right thing. The above decorator does not do the right thing. It does not deal with being invoked from multiple threads. If invoked in quick succession without having a cached result, it will pass all the calls through. A proper implementation would determine if an invoking thread is the first to trigger an update, stall all other calling threads while the update happens, and finally return the cached result to all waiting threads when it becomes available. That's not that hard to implement, but it's nowhere near the simplicity of not implementing it at all. The cost, I think, is much higher than explicitly exposing the polling interval. We could, of course, also go with a good enough implementation that doesn't do things quite right. I think that's just incurring technical debt. We'll end up with a method (status()) that seems OK to call freely, but it isn't quite so, and the problems arising from it would be subtle and hard to figure out once we forget about this little conversation. It would be like a button that says "press me", but which would, occasionally, make things fail in subtle ways. In other words, not a good idea. So anyway, I think we'd trade explicitness for hidden complexity. I'd rather have the former.

Even if we keep one loop (1), I don't think that's much different than (2). It would still involve a button that says "press me" which either fails in subtle ways or it's not worth making it function correctly.

Reply:
I understand. I'm working with the assumption that there's one central polling loop, which is why I haven't added any locking for thread safety to the decorator in the stale case. We should note in the docs that the status is not thread safe, and is not intended for the user to call when strategy is present.

I think 1 is reasonable and gets the job done.

To questions in 2:  
* assume one thread
* provider status polls block, so you can't call it again from 1 thread
* thread safety is not a good assumption in python anyway, so let's not add complexity until it's necessary. Adding a lock in the update method isn't hard.

Let's have a quick chat to make sure we are both on the same page.

Reply:
Here are some conclusions from an offline discussion: 
* decorator approach vs the PollItem are different implementations of the same functionality
* we want to keep the providers as simple and stateless as possible
* state/cache should be held in a centralized place and the PollItems do that.